streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -583,16 +583,19 @@ private void runLoop() {
                 }
             }
 
-            // try to process one fetch record from each task via the topology, and also trigger punctuate
+            // try to process several records from each task via the topology, and also trigger punctuate
             // functions if necessary, which may result in more records going through the topology in this loop
             if (totalNumBuffered > 0 || polledRecords) {
                 totalNumBuffered = 0;
 
                 if (!activeTasks.isEmpty()) {
                     for (StreamTask task : activeTasks.values()) {
-
-                        totalNumBuffered += task.process();
-
+                        // process this task's records to completion before
+                        // context switching to another task
+                        int numProcessed;
+                        while ((numProcessed = task.process()) != 0) {
+                            totalNumBuffered += numProcessed;
+                        }
                         requiresPoll = requiresPoll || task.requiresPoll();
 
                         streamsMetrics.processTimeSensor.record(computeLatency(), timerStartedMs);
```

Comment:
I'd like to do stats on a batch of requests, not each one individually anymore. The reason is that now that we do the processing in a tight loop, the performance of computeLatency (which calls time.milliseconds()) matters more than before (when perf was slower to start with). Basically either that or this sensor needs to be a DEBUG sensor.